qanta/buzzer/q_learning.py

```diff
@@ -0,0 +1,379 @@
+            actions = F.argmax(F.stack(qvalues), axis=2).data # length, batch
+            actions = actions.T.tolist()
+            for q, a in zip(batch.qids, actions):
+                q = q.tolist()
+                buzzes[q] = -1 if not any(a) else a.index(1)
+        return buzzes 
+    
+    def evaluate(self):
+
+        def _do_one(inputs):
+            t, action, result, length, hopeful, terminate = inputs
+        
+            if terminate or t >= length:
+                return 0, 0, 0, 0, 0, 0, True
+
+            reward, reward_hopeful = 0, 0
+            buzz, correct, rush, late = 0, 0, 0, 0
+            if action == 1:
+                terminate = True
+                buzz = 1
+                if result == 1:
+                    reward = 10
+                    correct = 1
+                else:
+                    reward = -5
+                    if hopeful:
+                        rush = 1
+                    #     reward -= 10
+            elif t == length - 1:
+                if hopeful:
+                    late = 1
+            #         reward = -10
+            reward_hopeful = reward if hopeful else 0
+
+            return reward, reward_hopeful, buzz, correct, rush, late, terminate
+
+        progress_bar = ProgressBar(self.eval_iter.size, unit_iteration=True)
+        epoch_stats = [0, 0, 0, 0, 0, 0]
+        num_examples = 0
+        num_hopeful = 0
+        for i in range(self.eval_iter.size):
+            batch = self.eval_iter.next_batch(self.model.xp)
+            length, batch_size, _ = batch.vecs.shape
+            qvalues = [self.model(vec) for vec in batch.vecs] # length * batch_size * 2
+            actions = F.argmax(F.stack(qvalues), axis=2).data # length * batch_size
+            hopeful = [any(x == 1) for x in batch.results.T] # batch_size
+            terminate = [False for _ in range(batch_size)]
+            num_examples += batch_size
+            num_hopeful += sum(hopeful)
+
+            for t in range(length):
+                tt = [t for _ in range(batch_size)]
+                inputs = zip(tt, actions[t], batch.results[t], batch.mask[t], hopeful, terminate)
+                returns = map(_do_one, inputs)
+                returns = list(map(list, zip(*returns)))
+                terminate = returns[-1]
+                returns = returns[:-1]
+                epoch_stats = list(map(lambda x, y: x + sum(y), epoch_stats, returns))
+                if all(terminate):
+                    break
+            # end of batch
+            progress_bar(*self.eval_iter.epoch_detail)
+        # end of epoch
+        self.eval_iter.finalize(reset=True)
+        progress_bar.finalize()
+        epoch_stats[0] /= num_examples # reward / num_total
+        epoch_stats[1] /= num_hopeful # reward_hopeful / num_hopeful
+        epoch_stats = [num_examples, num_hopeful] + epoch_stats
+        epoch_stats = BuzzStats(*epoch_stats)
+        return epoch_stats
+
+    # def train_one_epoch_new(self, progress_bar):
```

Comment:
Unless you have a good reason for keeping this around, its better to delete unused code and use git to recover it if we need it later